### PR TITLE
Export leanplum data to GCS

### PR DESCRIPTION
**Pull request by fbertsch**
There are directions for running locally if you want to give it a shot, but you'll need Leanplum access.

I'll run this in Airflow on GKE, similar to how we run `bigquery_etl` (note this is basically a mirror entrypoint file).

### CODE REVIEW

File: leanplum_data_export/export.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        return datatypes
+
+    def create_external_tables(self, bucket_name, prefix, date, tables, dataset):
+        gcs_loc = f"gs://{bucket_name}/{prefix}/{date}"
+
+        client = bigquery.Client()
+
+        dataset_ref = client.dataset(dataset)
+
+        for t in tables:
```
Comment (fbertsch) [inline leanplum_data_export/export.py:117]: Leanplum API returns files of the form:
`https://leanplum_export.storage.googleapis.com/export-$JOB_ID-$FILE_TYPE-$FILE_NO_OF_TYPE`

Where `$FILE_TYPE` is something like `outputsessions` and `outputexperiments`. They are all joinable to get e.g. events.

From there, we create tables of the form `$FILETYPE_$DATE`, which BQ handles well by default.

File: leanplum_data_export/export.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            datatypes |= set([datatype])
+
+            f = Path(local_filename)
+            base_dir = Path(datatype)
+            base_dir.mkdir(parents=True, exist_ok=True)
+
+            with requests.get(uri, stream=True) as r:
+                r.raise_for_status()
+                with f.open("wb") as opened:
+                    for chunk in r.iter_content(chunk_size=8192):
```
Comment (akkomar) [inline leanplum_data_export/export.py:97]: Why 8k chunks? Wouldn't this complete faster if larger chunks were used?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      - image: docker:18.02.0-ce
+    steps:
+      - setup_remote_docker
+      - restore_cache:
+          key: v1-{{.Branch}}
+      - run:
+          name: Restore Docker image cache
+          command: docker load -i /cache/docker.tar
+      - run:
+          name: Deploy to Dockerhub
```
Comment (akkomar) [inline .circleci/config.yml:70]: Why are we using public Dockerhub instead of private gcr repo here?

File: leanplum_data_export/export.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                        if chunk:
+                            opened.write(chunk)
```
Comment (akkomar) [inline leanplum_data_export/export.py:99]: TIL: this is equivalent to `shutil.copyfileobj(r.raw, opened)`

File: tests/test_leanplum_export.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+        with patch('leanplum_data_export.export.storage', spec=True) as MockStorage:
+            with pytest.raises(Exception) as e:
+                exporter.save_files(file_uris, bucket, prefix, date, "json")
+
+    @responses.activate
+    def test_export(self, exporter):
+        date = "20190101"
+        job_id = "testjobid"
+        file_uris = ['https://leanplum_export.storage.googleapis.com/export-5094741967896576-60c43e66-30fe-4e21-9bbd-563d2749b96f-outputsessions-0']
```
Comment (akkomar) [inline tests/test_leanplum_export.py:208]: Do these tests depend on files in `gs://leanplum_export`?

**Review by akkomar — COMMENTED** [review]
Added some comments. Mostly nits, the one about tests is the most important.

File: leanplum_data_export/export.py (hunk 2 above)
Comment (fbertsch) [reply leanplum_data_export/export.py:97]: Maybe? There's definitely a tradeoff between larger chunks and higher error rates.

File: .circleci/config.yml (hunk 3 above)
Comment (fbertsch) [reply .circleci/config.yml:70]: This is a conversation we had with ops last week. They didn't have a strong opinion either way. This is how we deploy all of our other jobs.

File: tests/test_leanplum_export.py (hunk 5 above)
Comment (fbertsch) [reply tests/test_leanplum_export.py:208]: Nope, no outside dependencies in these tests.

**Review by akkomar — APPROVED** [review]

**Comment by fbertsch** [comment]
Thanks for the review @akkomar!

**Pull request merged**